A40-csds-support.md

```diff
@@ -0,0 +1,328 @@
+xDS Configuration Dump via Client Status Discovery Service in gRPC
+----
+* Author(s): lidizheng
+* Approver: markdroth
+* Status: In-Review
+* Implemented in: All languages
+* Last updated: 2021-02-19
+* Discussion at: https://groups.google.com/g/grpc-io/c/zL45YyxtJ08
+
+## Abstract
+
+[Client Status Discovery
+Service](https://github.com/envoyproxy/envoy/blob/main/api/envoy/service/status/v3/csds.proto)
+(CSDS) is a service that exposes xDS config of a given client. It’s commonly
+used to query control planes for the synced xDS config of a particular sidecar
+proxy. However, it can also be used to query an xDS-compliant application for
+its received xDS configuration. This doc proposes a solution to implement a CSDS
+service in gRPC, so our users can debug their service mesh easily.
+
+
+## Background
+
+Envoy [started](https://github.com/envoyproxy/envoy/pull/9383) the CSDS
+development process in Dec 2019. The CSDS API has been available since xDS v2,
+and it’s under active development. But, Envoy proxies do not support or
+understand this service, only the control plane does. Envoy already provides
+config dump via its [admin
+interface](https://www.envoyproxy.io/docs/envoy/latest/operations/admin) for
+years. There is no visible plan for Envoy to support serving CSDS directly as a
+proxy.
+
+The CSDS service accepts two methods, one for unary, one for streaming. The
+expected behavior of the two methods is the same, but the streaming method can
+reuse the stream. According to the service description below, this is not a
+watch-style API.
+
+```proto
+// CSDS is Client Status Discovery Service. It can be used to get the status of
+// an xDS-compliant client from the management server's point of view. It can
+// also be used to get the current xDS states directly from the client.
+service ClientStatusDiscoveryService {
+  rpc StreamClientStatus(stream ClientStatusRequest) returns (stream ClientStatusResponse) {
+  }
+
+  rpc FetchClientStatus(ClientStatusRequest) returns (ClientStatusResponse) {
+  }
+}
+```
+
+### Related Proposals: 
+* [A27 - xDS-Based Global Load
+  Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+* [A38 - Admin Interface API](https://github.com/grpc/proposal/pull/218)
+
+
+## Proposal
+
+### CSDS in gRPC
+
+The goal of adding a CSDS service to gRPC is to enable programmatic access to
+the operating xDS configs of a running gRPC application. To make this more
+clear, an xDS-compliant application receives many xDS configs, which may be
+rejected or ignored or obsoleted. **The gRPC CSDS service should always return
+the currently accepted xDS configs.**
+
+### Config Status in gRPC
```

Comment:
The new `ClientResourceStatus` enum has already been accepted in the xDS protos, so we don't need to include it in this gRFC.  That's now part of the CSDS API, and we're just implementing that API.

Reply:
This section is removed.